rust/gitxetcore/src/config/permission.rs

```diff
@@ -37,30 +54,125 @@ impl Permission {
+            return Ok(path.to_owned());
+        }
+
+        info!("Lack permission for R+W into {path:?}");
+
+        if suggest {
+            if let Some(suggest_list) = priority_suggest_list {
+                for &path in suggest_list {
+                    if may_have_write_permission_into(path, expect_type) {
+                        return Ok(path.to_owned());
+                    }
+                    info!("Lack permission for R+W into {path:?}");
+                }
+            }
+
+            // Cannot write into path, try append the path by "_[username]"
+            let mut path = path.to_owned();
+            let pathstr = path.as_mut_os_string();
+            pathstr.push(format!("_{}", whoami::username()));
+
+            if may_have_write_permission_into(&path, expect_type) {
+                return Ok(path);
+            }
+
+            info!("Lack permission for R+W into {path:?}");
+
+            let last_component = format!(
+                "{}_{}",
+                suggest_prefix.unwrap_or_default(),
+                thread_rng().gen::<u32>() // thread_rng is CSPRNG, very unlikely to collide
+            );
+
+            // Cannot write into "path_[username]", try a random path "[prefix]_[xxx]" under HOME
+            if let Some(home) = dirs::home_dir() {
+                let path = home.join(&last_component);
+
+                if may_have_write_permission_into(&path, expect_type) {
+                    return Ok(path);
+                }
+
+                info!("Lack permission for R+W into {path:?}");
+            }
+
+            // Cannot write into home directory, try a random path "[prefix]_[xxx]" under /tmp
+            let path = std::env::temp_dir().join(&last_component);
 
-            eprintln!("{}", message.bright_blue());
-            warn!("Xet directory {path:?} created with elevated privileges");
+            if may_have_write_permission_into(&path, expect_type) {
+                return Ok(path);
+            }
+
+            info!("Lack permission for R+W into {path:?}");
+
+            // Cannot write into tmp directory, try a random path "[prefix]_[xxx]" under the cwd
+            let path = std::env::current_dir()
+                .unwrap_or_default()
+                .join(&last_component);
+
+            if may_have_write_permission_into(&path, expect_type) {
+                return Ok(path);
+            }
+
+            info!("Lack permission for R+W into {path:?}");
         }
+
+        Err(anyhow!("Fail to find a path with correct permission"))
     }
 }
 
-/// Checks if the program is run under elevated privilege
-
+/// Checks if the program is running under elevated privilege
 fn is_elevated() -> bool {
     // In a Unix-like environment, when a program is run with sudo,
     // the effective user ID (euid) of the process is set to 0.
     #[cfg(unix)]
-    return unsafe { libc::geteuid() == 0 };
+    {
+        unsafe { libc::getegid() == 0 }
```

Comment:
This switches to group id?

Reply:
The "effective user ID", which is set to 0 when running with 'sudo'. (Commented above)

Reply:
Noting that it is getegid, which is the group id

Reply:
Ah good catch!